core/src/main/java/org/infinispan/remoting/transport/jgroups/StaggeredRequest.java

```diff
@@ -95,8 +95,12 @@ void sendNextMessage() {
             isFinalTarget = targetIndex >= getTargetsSize();
          }
 
-         // Sending may block in flow-control or even in TCP, so we must do it outside the critical section
-         transport.sendCommand(target, command, requestId, deliverOrder, true, false);
+         Address finalTarget = target;
+         transport.blockingManager.runBlocking(() -> {
```

Comment:
My original idea was to use a Flowable per target so that way we would only block a single thread per target at worst case and if needed this allows for concurrency safeties by target address if that helps.

Reply:
I don't think we want to use a single thread per target. 
If you have 2 concurrent read requests (different keys) targetting the same destination, one key with 10 bytes and another with 10 kbytes, you could have enough credits to let the small key read go through and only block the large key request.

In any case, I'm expecting the perf for this change to be worst because I think, in most cases, we do not block anywhere, and the penalty of moving to a different thread will show in the results.

Reply:
Just for informational purposes: Does Jgroups allow barging of credits like that? If so that could cause a livelock if credits are not replenished in a bigger block than the request size. So I guess FRAG would have to be smaller than the credit replenishes.